File: backend/src/services/taxService.js
// src/services/taxService.js
// Impostos aplicáveis (lei angolana):
//  - IVA: 14% sobre tudo (produtos e serviços). Soma-se à fatura (o comprador
//    paga o total com IVA).
//  - Retenção na Fonte de Imposto Industrial (Lei 26/20): 6,5% sobre SERVIÇOS.
//    NÃO soma à fatura — o comprador desconta 6,5% do valor do serviço e
//    entrega à AGT por conta do fornecedor. Reduz o líquido do fornecedor.
// Ambas as taxas são configuráveis por ambiente.
const IVA_RATE = Number(process.env.IVA_RATE) || 0.14;
const WITHHOLDING_RATE = Number(process.env.WITHHOLDING_RATE) || 0.065; // retenção II (serviços)

const round2 = (n) => Math.round((Number(n) + Number.EPSILON) * 100) / 100;

// O IVA é igual para produto e serviço; `kind` é aceite por compatibilidade.
function rateForKind() {
  return IVA_RATE;
}

// IVA de um montante líquido.
function computeTax(netAmount) {
  const tax = round2(Number(netAmount) * IVA_RATE);
  return { rate: IVA_RATE, net: round2(netAmount), tax, gross: round2(Number(netAmount) + tax) };
}

// Retenção na fonte (II) de uma linha — só se aplica a serviços.
function withholdingFor(netAmount, kind) {
  return kind === 'SERVICO' ? round2(Number(netAmount) * WITHHOLDING_RATE) : 0;
}

// Agrega várias linhas ({ net, kind }) com IVA (14% em tudo) e retenção na
// fonte (6,5% só nas linhas de serviço).
//  - gross    = net + IVA          (total a pagar pelo comprador)
//  - withheld = retenção sobre serviços (descontada ao fornecedor)
//  - supplierNet = gross - withheld (líquido que o fornecedor recebe)
function summarize(lines) {
  const net = round2(lines.reduce((s, l) => s + Number(l.net || 0), 0));
  const tax = round2(net * IVA_RATE);
  const withheld = round2(lines.reduce((s, l) => s + withholdingFor(l.net, l.kind), 0));
  const gross = round2(net + tax);
  return { net, tax, gross, withheld, supplierNet: round2(gross - withheld), rate: IVA_RATE, withholdingRate: WITHHOLDING_RATE };
}

module.exports = { IVA_RATE, WITHHOLDING_RATE, rateForKind, computeTax, withholdingFor, summarize };

File: backend-java/src/main/java/ao/kixima/tax/TaxService.java
package ao.kixima.tax;

import ao.kixima.catalog.ProductKind;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.List;

/**
 * Espelha backend/src/services/taxService.js — impostos da lei angolana:
 * IVA 14% sobre tudo, retenção na fonte (II, Lei 26/20) 6,5% só sobre
 * SERVIÇOS, reduzindo o líquido do fornecedor (não soma à fatura). O
 * arredondamento de taxService.js é sempre ao cêntimo mais próximo
 * ({@code Math.round}), nunca por excesso (essa é uma regra à parte, só
 * para o payload AGT — ver faturacaoService.arredondarPorExcessoAoCentimo,
 * M4) — por isso HALF_UP em vez de CEILING aqui.
 */
@Service
public class TaxService {

    private final BigDecimal ivaRate;
    private final BigDecimal withholdingRate;

    public TaxService(@Value("${kixima.tax.iva-rate:0.14}") String ivaRate,
                       @Value("${kixima.tax.withholding-rate:0.065}") String withholdingRate) {
        this.ivaRate = new BigDecimal(ivaRate);
        this.withholdingRate = new BigDecimal(withholdingRate);
    }

    public BigDecimal getIvaRate() {
        return ivaRate;
    }

    private BigDecimal round2(BigDecimal n) {
        return n.setScale(2, RoundingMode.HALF_UP);
    }

    public record ComputedTax(BigDecimal rate, BigDecimal net, BigDecimal tax, BigDecimal gross) {
    }

    public ComputedTax computeTax(BigDecimal netAmount) {
        BigDecimal net = round2(netAmount);
        BigDecimal tax = round2(netAmount.multiply(ivaRate));
        return new ComputedTax(ivaRate, net, tax, round2(netAmount.add(tax)));
    }

    /** Retenção na fonte (II) de uma linha — só se aplica a serviços. */
    public BigDecimal withholdingFor(BigDecimal netAmount, ProductKind kind) {
        return kind == ProductKind.SERVICO ? round2(netAmount.multiply(withholdingRate)) : BigDecimal.ZERO.setScale(2);
    }

    public record Line(BigDecimal net, ProductKind kind) {
    }

    public record Summary(BigDecimal net, BigDecimal tax, BigDecimal gross, BigDecimal withheld,
                           BigDecimal supplierNet, BigDecimal rate, BigDecimal withholdingRate) {
    }

    /**
     * Agrega várias linhas com IVA (14% em tudo) e retenção na fonte (6,5%
     * só nas linhas de serviço). gross = net + IVA (total a pagar pelo
     * comprador); supplierNet = gross - withheld (líquido do fornecedor).
     */
    public Summary summarize(List<Line> lines) {
        BigDecimal net = round2(lines.stream().map(Line::net).reduce(BigDecimal.ZERO, BigDecimal::add));
        BigDecimal tax = round2(net.multiply(ivaRate));
        BigDecimal withheld = round2(lines.stream()
                .map(l -> withholdingFor(l.net(), l.kind()))
                .reduce(BigDecimal.ZERO, BigDecimal::add));
        BigDecimal gross = round2(net.add(tax));
        return new Summary(net, tax, gross, withheld, round2(gross.subtract(withheld)), ivaRate, withholdingRate);
    }
}
